crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -2340,11 +2347,50 @@ impl<TX: DbTxMut + DbTx + 'static, N: NodeTypes> TrieWriter for DatabaseProvider
 
         Ok(num_entries)
     }
+
+    /// Records the current values of all trie nodes which will be updated using the [`TrieUpdates`]
+    /// into the trie changesets tables.
+    ///
+    /// The intended usage of this method is to call it _prior_ to calling `write_trie_updates` with
+    /// the same [`TrieUpdates`].
+    ///
+    /// Returns the number of keys written.
+    fn write_trie_changesets(
```

Comment:
I wonder if we should combine this just into `write_trie_updates` method for two reasons:
1. We will not need to be careful with calling `write_trie_updates` before `write_trie_updates`, because it will all be encapsulated in the same method
2. We can update `AccountsTrie` and `StoragesTrie` at the same time as writing changesets

Reply:
I _think_ it will be necessary to have a separate `write_trie_changesets` style method for #18464, though I'm not 100% this is the exact form it will take. But we could have the method be called internal to `write_trie_updates` automatically, but still expose it through the trait?